libraries/chem-meta/src/parsing-utils/molfile-handler.ts

```diff
@@ -0,0 +1,163 @@
+      this.shiftIdxToAtomTypeV3K;
+
+    this.getBondBlockIdx = isV2K ? this.getBondBlockIdxV2K : this.getBondBlockIdxV3K;
+  }
+
+  public static createInstance(file: string): MolfileHandler {
+    if (!this.instance)
+      this.instance = new MolfileHandler(file); // a workaround to define an abstract singleton
+    return this.instance as MolfileHandler;
+  }
+
+  // Protected members
+
+  protected parseAtomAndBondCounts!: () => AtomAndBondCounts;
+  protected getCountsLineIdx!: () => number;
+  protected getAtomBlockIdx!: () => number;
+  protected shiftIdxToXColumn!: (lineStartIdx: number) => number;
+  protected shiftIdxToAtomType!: (idx: number) => number;
+  protected getBondBlockIdx!: () => number;
+
+  protected parseAtomType(idx: number): string {
+    const begin = idx;
+    const end = this.file.indexOf(' ', begin);
+    return this.file.substring(begin, end);
+  }
+
+  // Private members
+
+  /** Determine whether the file is V2000/V3000, or throw */
+  private static determineMolfileVersion(molfile: string): MOLFILE_VERSION {
+    if (MolfileHandler.validateV3K(molfile))
+      return MOLFILE_VERSION.V3000;
+    else if (MolfileHandler.validateV2K(molfile))
+      return MOLFILE_VERSION.V2000;
+    else
+      throw new Error('Malformed molfile');
+  }
+
+  private shiftIdxToAtomTypeV2K(lineStartIdx: number): number {
+    return this.shiftToSpecifiedColumn(lineStartIdx, 4);
+  }
+
+  private shiftIdxToAtomTypeV3K(lineStartIdx: number): number {
+    return this.shiftToSpecifiedColumn(lineStartIdx, 3);
+  }
+
+  private getCountsLineV2KIdx(): number {
+    let idx = 0;
+    // 3 is for the number of header lines in molfiles
+    for (let i = 0; i < 3; ++i)
+      idx = this.getNextLineIdx(idx);
+    return idx;
+  }
+
+  private getCountsLineV3KIdx(): number {
+    return this.file.indexOf(V3K.BEGIN_COUNTS_LINE);
+  }
+
+  private getAtomBlockIdxV2K(): number {
+    let idx = this.getCountsLineIdx();
+    idx = this.getNextLineIdx(idx);
+    return idx;
+  }
+
+  private getAtomBlockIdxV3K(): number {
+    let idx = this.file.indexOf(V3K.BEGIN_ATOM_BLOCK);
+    idx = this.getNextLineIdx(idx);
+    return idx;
+  }
+
+  private shiftIdxToXColumnV2K(lineStartIdx: number): number {
+    return this.getNextColumnIdx(lineStartIdx);
+  }
+
+  private shiftIdxToXColumnV3K(lineStartIdx: number): number {
+    return this.shiftToSpecifiedColumn(lineStartIdx, 4);
+  }
+
+  private getBondBlockIdxV2K(): number {
+    let idx = this.getAtomBlockIdx();
+    for (let i = 0; i < this.atomCount; i++)
+      idx = this.getNextLineIdx(idx);
+    return idx;
+  }
+
+  private getBondBlockIdxV3K(): number {
+    return this.getNextLineIdx(this.file.indexOf(V3K.BEGIN_BOND_BLOCK));
+  }
+
+  // todo: devise a more reliable validation
+  private static validateV3K(molfile: string): boolean {
+    if (molfile.indexOf(V3K.HEADER) !== -1 &&
+    molfile.indexOf(V3K.END) !== -1)
+      return true;
+    return false;
```

Comment:
you can rewrite it to a single
`return (molfile.indexOf(V3K.HEADER) !== -1 && molfile.indexOf(V3K.END) !== -1)`